### ISSUE TITLE
Single function for reading xyz and writing xyz files

### ISSUE THREAD

**Issue by NikoOinonen** [Probe-Particle/ProbeParticleModel#24]
In basUtils there are 4 different functions for reading xyz files that return results in slightly different formats and 6 different functions for writing xyz files. We should agree on single interface for these.

**Comment by ondrejkrejci** [comment]
Unless I am wrong, this is where we should look at:
https://github.com/Probe-Particle/ProbeParticleModel/blob/a97b04adcfea4228384c4bd946001599c34b4a83/pyProbeParticle/basUtils.py#L440 
Most of the important scripts (`generateLJFF.py` and `extractPauli.py`) are using it. We should go through other scripts in the _root_ to see, whether they are using anything from the other one; if not/if repaired, we can use `loadAtoms` only.

**Comment by ondrejkrejci** [comment]
#39 is kind related to this one. Generally it would be good to reorganize `basUtils.py` with having reading procedures in the top and writing ones in the bottom.

**Comment by NikoOinonen** [comment]
`loadGeometry` seems to be more an ease-of-use function for automatically dealing with all different formats. I'm more thinking that we should have just one function called `loadXYZ` and another called `saveXYZ` for reading and writing xyz files (which `loadGeometry` would use internally, too). For `loadXYZ` I would suggest that it returns a tuple of numpy arrays `(xyzs, Zs, qs)` for the coordinates, atomic numbers, and point-charges of the atoms.

Right now for reading xyz files we have in `basUtils.py`: `loadBas`, `loadAtoms`, `loadAtomsNP`, and `loadAtomsLines`.
Additionally, for some reason in `atomicUtils.py` there are exact duplicates of `loadAtoms` and `loadAtomsNP`.

And for writing xyz files we have in `basUtils.py`: `writeAtoms`, `writeAtomsTransposed`, `saveXyz_Transposed`, `saveXyz`, `writeDebugXYZ`, `writeDebugXYZ_2`, `writeDebugXYZ__`, and in `atomicUtils.py` there is `saveAtoms`, `writeToXYZ`, `saveXYZ`, as well as `writeToXYZ` in `SurfConf.py`.

Basically, I would go over every location where these are in use within the repo and replace them with the new functions. I could do this at some point.

**Comment by ondrejkrejci** [comment]
About writing XYZ: We need one of them for the STM part - I will specify on it.

**Comment by ondrejkrejci** [comment]
An xyz.example, where you have specified lattice, created through ASE. The problem here is that the fifth column are not charges, but just layer information. But we cannot make thic code 100% stupid proof.
```
400
Lattice="40.587929240107826 0.0 0.0 0.0 35.15017780893861 0.0 0.0 0.0 42.485492908861346" Properties=species:S:1:pos:R:3:tags:I:1 pbc="T T T"
Fe       0.00000000       0.00000000      20.00000000        4
Fe       4.05879292       0.00000000      20.00000000        4
Fe       8.11758585       0.00000000      20.00000000        4
Fe      12.17637877       0.00000000      20.00000000        4
Fe      16.23517170       0.00000000      20.00000000        4
Fe      20.29396462       0.00000000      20.00000000        4
Fe      24.35275754       0.00000000      20.00000000        4
Fe      28.41155047       0.00000000      20.00000000        4
Fe      32.47034339       0.00000000      20.00000000        4
Fe      36.52913632       0.00000000      20.00000000        4
Fe       2.02939646       3.51501778      20.00000000        4
Fe       6.08818939       3.51501778      20.00000000        4
Fe      10.14698231       3.51501778      20.00000000        4
Fe      14.20577523       3.51501778      20.00000000        4
Fe      18.26456816       3.51501778      20.00000000        4
Fe      22.32336108       3.51501778      20.00000000        4
Fe      26.38215401       3.51501778      20.00000000        4
Fe      30.44094693       3.51501778      20.00000000        4
Fe      34.49973985       3.51501778      20.00000000        4
Fe      38.55853278       3.51501778      20.00000000        4
Fe       0.00000000       7.03003556      20.00000000        4
Fe       4.05879292       7.03003556      20.00000000        4
Fe       8.11758585       7.03003556      20.00000000        4
Fe      12.17637877       7.03003556      20.00000000        4
Fe      16.23517170       7.03003556      20.00000000        4
Fe      20.29396462       7.03003556      20.00000000        4
Fe      24.35275754       7.03003556      20.00000000        4
Fe      28.41155047       7.03003556      20.00000000        4
Fe      32.47034339       7.03003556      20.00000000        4
Fe      36.52913632       7.03003556      20.00000000        4
Fe       2.02939646      10.54505334      20.00000000        4
Fe       6.08818939      10.54505334      20.00000000        4
Fe      10.14698231      10.54505334      20.00000000        4
Fe      14.20577523      10.54505334      20.00000000        4
Fe      18.26456816      10.54505334      20.00000000        4
Fe      22.32336108      10.54505334      20.00000000        4
Fe      26.38215401      10.54505334      20.00000000        4
Fe      30.44094693      10.54505334      20.00000000        4
Fe      34.49973985      10.54505334      20.00000000        4
Fe      38.55853278      10.54505334      20.00000000        4
Fe       0.00000000      14.06007112      20.00000000        4
Fe       4.05879292      14.06007112      20.00000000        4
Fe       8.11758585      14.06007112      20.00000000        4
Fe      12.17637877      14.06007112      20.00000000        4
Fe      16.23517170      14.06007112      20.00000000        4
Fe      20.29396462      14.06007112      20.00000000        4
Fe      24.35275754      14.06007112      20.00000000        4
Fe      28.41155047      14.06007112      20.00000000        4
Fe      32.47034339      14.06007112      20.00000000        4
Fe      36.52913632      14.06007112      20.00000000        4
Fe       2.02939646      17.57508890      20.00000000        4
Fe       6.08818939      17.57508890      20.00000000        4
Fe      10.14698231      17.57508890      20.00000000        4
Fe      14.20577523      17.57508890      20.00000000        4
Fe      18.26456816      17.57508890      20.00000000        4
Fe      22.32336108      17.57508890      20.00000000        4
Fe      26.38215401      17.57508890      20.00000000        4
Fe      30.44094693      17.57508890      20.00000000        4
Fe      34.49973985      17.57508890      20.00000000        4
Fe      38.55853278      17.57508890      20.00000000        4
Fe       0.00000000      21.09010669      20.00000000        4
Fe       4.05879292      21.09010669      20.00000000        4
Fe       8.11758585      21.09010669      20.00000000        4
Fe      12.17637877      21.09010669      20.00000000        4
Fe      16.23517170      21.09010669      20.00000000        4
Fe      20.29396462      21.09010669      20.00000000        4
Fe      24.35275754      21.09010669      20.00000000        4
Fe      28.41155047      21.09010669      20.00000000        4
Fe      32.47034339      21.09010669      20.00000000        4
Fe      36.52913632      21.09010669      20.00000000        4
Fe       2.02939646      24.60512447      20.00000000        4
Fe       6.08818939      24.60512447      20.00000000        4
Fe      10.14698231      24.60512447      20.00000000        4
Fe      14.20577523      24.60512447      20.00000000        4
Fe      18.26456816      24.60512447      20.00000000        4
Fe      22.32336108      24.60512447      20.00000000        4
Fe      26.38215401      24.60512447      20.00000000        4
Fe      30.44094693      24.60512447      20.00000000        4
Fe      34.49973985      24.60512447      20.00000000        4
Fe      38.55853278      24.60512447      20.00000000        4
Fe       0.00000000      28.12014225      20.00000000        4
Fe       4.05879292      28.12014225      20.00000000        4
Fe       8.11758585      28.12014225      20.00000000        4
Fe      12.17637877      28.12014225      20.00000000        4
Fe      16.23517170      28.12014225      20.00000000        4
Fe      20.29396462      28.12014225      20.00000000        4
Fe      24.35275754      28.12014225      20.00000000        4
Fe      28.41155047      28.12014225      20.00000000        4
Fe      32.47034339      28.12014225      20.00000000        4
Fe      36.52913632      28.12014225      20.00000000        4
Fe       2.02939646      31.63516003      20.00000000        4
Fe       6.08818939      31.63516003      20.00000000        4
Fe      10.14698231      31.63516003      20.00000000        4
Fe      14.20577523      31.63516003      20.00000000        4
Fe      18.26456816      31.63516003      20.00000000        4
Fe      22.32336108      31.63516003      20.00000000        4
Fe      26.38215401      31.63516003      20.00000000        4
Fe      30.44094693      31.63516003      20.00000000        4
Fe      34.49973985      31.63516003      20.00000000        4
Fe      38.55853278      31.63516003      20.00000000        4
Fe       2.02939646       1.17167259      20.82849764        3
Fe       6.08818939       1.17167259      20.82849764        3
Fe      10.14698231       1.17167259      20.82849764        3
Fe      14.20577523       1.17167259      20.82849764        3
Fe      18.26456816       1.17167259      20.82849764        3
Fe      22.32336108       1.17167259      20.82849764        3
Fe      26.38215401       1.17167259      20.82849764        3
Fe      30.44094693       1.17167259      20.82849764        3
Fe      34.49973985       1.17167259      20.82849764        3
Fe      38.55853278       1.17167259      20.82849764        3
Fe       0.00000000       4.68669037      20.82849764        3
Fe       4.05879292       4.68669037      20.82849764        3
Fe       8.11758585       4.68669037      20.82849764        3
Fe      12.17637877       4.68669037      20.82849764        3
Fe      16.23517170       4.68669037      20.82849764        3
Fe      20.29396462       4.68669037      20.82849764        3
Fe      24.35275754       4.68669037      20.82849764        3
Fe      28.41155047       4.68669037      20.82849764        3
Fe      32.47034339       4.68669037      20.82849764        3
Fe      36.52913632       4.68669037      20.82849764        3
Fe       2.02939646       8.20170816      20.82849764        3
Fe       6.08818939       8.20170816      20.82849764        3
Fe      10.14698231       8.20170816      20.82849764        3
Fe      14.20577523       8.20170816      20.82849764        3
Fe      18.26456816       8.20170816      20.82849764        3
Fe      22.32336108       8.20170816      20.82849764        3
Fe      26.38215401       8.20170816      20.82849764        3
Fe      30.44094693       8.20170816      20.82849764        3
Fe      34.49973985       8.20170816      20.82849764        3
Fe      38.55853278       8.20170816      20.82849764        3
Fe       0.00000000      11.71672594      20.82849764        3
Fe       4.05879292      11.71672594      20.82849764        3
Fe       8.11758585      11.71672594      20.82849764        3
Fe      12.17637877      11.71672594      20.82849764        3
Fe      16.23517170      11.71672594      20.82849764        3
Fe      20.29396462      11.71672594      20.82849764        3
Fe      24.35275754      11.71672594      20.82849764        3
Fe      28.41155047      11.71672594      20.82849764        3
Fe      32.47034339      11.71672594      20.82849764        3
Fe      36.52913632      11.71672594      20.82849764        3
Fe       2.02939646      15.23174372      20.82849764        3
Fe       6.08818939      15.23174372      20.82849764        3
Fe      10.14698231      15.23174372      20.82849764        3
Fe      14.20577523      15.23174372      20.82849764        3
Fe      18.26456816      15.23174372      20.82849764        3
Fe      22.32336108      15.23174372      20.82849764        3
Fe      26.38215401      15.23174372      20.82849764        3
Fe      30.44094693      15.23174372      20.82849764        3
Fe      34.49973985      15.23174372      20.82849764        3
Fe      38.55853278      15.23174372      20.82849764        3
Fe       0.00000000      18.74676150      20.82849764        3
Fe       4.05879292      18.74676150      20.82849764        3
Fe       8.11758585      18.74676150      20.82849764        3
Fe      12.17637877      18.74676150      20.82849764        3
Fe      16.23517170      18.74676150      20.82849764        3
Fe      20.29396462      18.74676150      20.82849764        3
Fe      24.35275754      18.74676150      20.82849764        3
Fe      28.41155047      18.74676150      20.82849764        3
Fe      32.47034339      18.74676150      20.82849764        3
Fe      36.52913632      18.74676150      20.82849764        3
Fe       2.02939646      22.26177928      20.82849764        3
Fe       6.08818939      22.26177928      20.82849764        3
Fe      10.14698231      22.26177928      20.82849764        3
Fe      14.20577523      22.26177928      20.82849764        3
Fe      18.26456816      22.26177928      20.82849764        3
Fe      22.32336108      22.26177928      20.82849764        3
Fe      26.38215401      22.26177928      20.82849764        3
Fe      30.44094693      22.26177928      20.82849764        3
Fe      34.49973985      22.26177928      20.82849764        3
Fe      38.55853278      22.26177928      20.82849764        3
Fe       0.00000000      25.77679706      20.82849764        3
Fe       4.05879292      25.77679706      20.82849764        3
Fe       8.11758585      25.77679706      20.82849764        3
Fe      12.17637877      25.77679706      20.82849764        3
Fe      16.23517170      25.77679706      20.82849764        3
Fe      20.29396462      25.77679706      20.82849764        3
Fe      24.35275754      25.77679706      20.82849764        3
Fe      28.41155047      25.77679706      20.82849764        3
Fe      32.47034339      25.77679706      20.82849764        3
Fe      36.52913632      25.77679706      20.82849764        3
Fe       2.02939646      29.29181484      20.82849764        3
Fe       6.08818939      29.29181484      20.82849764        3
Fe      10.14698231      29.29181484      20.82849764        3
Fe      14.20577523      29.29181484      20.82849764        3
Fe      18.26456816      29.29181484      20.82849764        3
Fe      22.32336108      29.29181484      20.82849764        3
Fe      26.38215401      29.29181484      20.82849764        3
Fe      30.44094693      29.29181484      20.82849764        3
Fe      34.49973985      29.29181484      20.82849764        3
Fe      38.55853278      29.29181484      20.82849764        3
Fe       0.00000000      32.80683262      20.82849764        3
Fe       4.05879292      32.80683262      20.82849764        3
Fe       8.11758585      32.80683262      20.82849764        3
Fe      12.17637877      32.80683262      20.82849764        3
Fe      16.23517170      32.80683262      20.82849764        3
Fe      20.29396462      32.80683262      20.82849764        3
Fe      24.35275754      32.80683262      20.82849764        3
Fe      28.41155047      32.80683262      20.82849764        3
Fe      32.47034339      32.80683262      20.82849764        3
Fe      36.52913632      32.80683262      20.82849764        3
Fe       0.00000000       2.34334519      21.65699527        2
Fe       4.05879292       2.34334519      21.65699527        2
Fe       8.11758585       2.34334519      21.65699527        2
Fe      12.17637877       2.34334519      21.65699527        2
Fe      16.23517170       2.34334519      21.65699527        2
Fe      20.29396462       2.34334519      21.65699527        2
Fe      24.35275754       2.34334519      21.65699527        2
Fe      28.41155047       2.34334519      21.65699527        2
Fe      32.47034339       2.34334519      21.65699527        2
Fe      36.52913632       2.34334519      21.65699527        2
Fe       2.02939646       5.85836297      21.65699527        2
Fe       6.08818939       5.85836297      21.65699527        2
Fe      10.14698231       5.85836297      21.65699527        2
Fe      14.20577523       5.85836297      21.65699527        2
Fe      18.26456816       5.85836297      21.65699527        2
Fe      22.32336108       5.85836297      21.65699527        2
Fe      26.38215401       5.85836297      21.65699527        2
Fe      30.44094693       5.85836297      21.65699527        2
Fe      34.49973985       5.85836297      21.65699527        2
Fe      38.55853278       5.85836297      21.65699527        2
Fe       0.00000000       9.37338075      21.65699527        2
Fe       4.05879292       9.37338075      21.65699527        2
Fe       8.11758585       9.37338075      21.65699527        2
Fe      12.17637877       9.37338075      21.65699527        2
Fe      16.23517170       9.37338075      21.65699527        2
Fe      20.29396462       9.37338075      21.65699527        2
Fe      24.35275754       9.37338075      21.65699527        2
Fe      28.41155047       9.37338075      21.65699527        2
Fe      32.47034339       9.37338075      21.65699527        2
Fe      36.52913632       9.37338075      21.65699527        2
Fe       2.02939646      12.88839853      21.65699527        2
Fe       6.08818939      12.88839853      21.65699527        2
Fe      10.14698231      12.88839853      21.65699527        2
Fe      14.20577523      12.88839853      21.65699527        2
Fe      18.26456816      12.88839853      21.65699527        2
Fe      22.32336108      12.88839853      21.65699527        2
Fe      26.38215401      12.88839853      21.65699527        2
Fe      30.44094693      12.88839853      21.65699527        2
Fe      34.49973985      12.88839853      21.65699527        2
Fe      38.55853278      12.88839853      21.65699527        2
Fe       0.00000000      16.40341631      21.65699527        2
Fe       4.05879292      16.40341631      21.65699527        2
Fe       8.11758585      16.40341631      21.65699527        2
Fe      12.17637877      16.40341631      21.65699527        2
Fe      16.23517170      16.40341631      21.65699527        2
Fe      20.29396462      16.40341631      21.65699527        2
Fe      24.35275754      16.40341631      21.65699527        2
Fe      28.41155047      16.40341631      21.65699527        2
Fe      32.47034339      16.40341631      21.65699527        2
Fe      36.52913632      16.40341631      21.65699527        2
Fe       2.02939646      19.91843409      21.65699527        2
Fe       6.08818939      19.91843409      21.65699527        2
Fe      10.14698231      19.91843409      21.65699527        2
Fe      14.20577523      19.91843409      21.65699527        2
Fe      18.26456816      19.91843409      21.65699527        2
Fe      22.32336108      19.91843409      21.65699527        2
Fe      26.38215401      19.91843409      21.65699527        2
Fe      30.44094693      19.91843409      21.65699527        2
Fe      34.49973985      19.91843409      21.65699527        2
Fe      38.55853278      19.91843409      21.65699527        2
Fe       0.00000000      23.43345187      21.65699527        2
Fe       4.05879292      23.43345187      21.65699527        2
Fe       8.11758585      23.43345187      21.65699527        2
Fe      12.17637877      23.43345187      21.65699527        2
Fe      16.23517170      23.43345187      21.65699527        2
Fe      20.29396462      23.43345187      21.65699527        2
Fe      24.35275754      23.43345187      21.65699527        2
Fe      28.41155047      23.43345187      21.65699527        2
Fe      32.47034339      23.43345187      21.65699527        2
Fe      36.52913632      23.43345187      21.65699527        2
Fe       2.02939646      26.94846965      21.65699527        2
Fe       6.08818939      26.94846965      21.65699527        2
Fe      10.14698231      26.94846965      21.65699527        2
Fe      14.20577523      26.94846965      21.65699527        2
Fe      18.26456816      26.94846965      21.65699527        2
Fe      22.32336108      26.94846965      21.65699527        2
Fe      26.38215401      26.94846965      21.65699527        2
Fe      30.44094693      26.94846965      21.65699527        2
Fe      34.49973985      26.94846965      21.65699527        2
Fe      38.55853278      26.94846965      21.65699527        2
Fe       0.00000000      30.46348743      21.65699527        2
Fe       4.05879292      30.46348743      21.65699527        2
Fe       8.11758585      30.46348743      21.65699527        2
Fe      12.17637877      30.46348743      21.65699527        2
Fe      16.23517170      30.46348743      21.65699527        2
Fe      20.29396462      30.46348743      21.65699527        2
Fe      24.35275754      30.46348743      21.65699527        2
Fe      28.41155047      30.46348743      21.65699527        2
Fe      32.47034339      30.46348743      21.65699527        2
Fe      36.52913632      30.46348743      21.65699527        2
Fe       2.02939646      33.97850522      21.65699527        2
Fe       6.08818939      33.97850522      21.65699527        2
Fe      10.14698231      33.97850522      21.65699527        2
Fe      14.20577523      33.97850522      21.65699527        2
Fe      18.26456816      33.97850522      21.65699527        2
Fe      22.32336108      33.97850522      21.65699527        2
Fe      26.38215401      33.97850522      21.65699527        2
Fe      30.44094693      33.97850522      21.65699527        2
Fe      34.49973985      33.97850522      21.65699527        2
Fe      38.55853278      33.97850522      21.65699527        2
Fe       0.00000000       0.00000000      22.48549291        1
Fe       4.05879292       0.00000000      22.48549291        1
Fe       8.11758585       0.00000000      22.48549291        1
Fe      12.17637877       0.00000000      22.48549291        1
Fe      16.23517170       0.00000000      22.48549291        1
Fe      20.29396462       0.00000000      22.48549291        1
Fe      24.35275754       0.00000000      22.48549291        1
Fe      28.41155047       0.00000000      22.48549291        1
Fe      32.47034339       0.00000000      22.48549291        1
Fe      36.52913632       0.00000000      22.48549291        1
Fe       2.02939646       3.51501778      22.48549291        1
Fe       6.08818939       3.51501778      22.48549291        1
Fe      10.14698231       3.51501778      22.48549291        1
Fe      14.20577523       3.51501778      22.48549291        1
Fe      18.26456816       3.51501778      22.48549291        1
Fe      22.32336108       3.51501778      22.48549291        1
Fe      26.38215401       3.51501778      22.48549291        1
Fe      30.44094693       3.51501778      22.48549291        1
Fe      34.49973985       3.51501778      22.48549291        1
Fe      38.55853278       3.51501778      22.48549291        1
Fe       0.00000000       7.03003556      22.48549291        1
Fe       4.05879292       7.03003556      22.48549291        1
Fe       8.11758585       7.03003556      22.48549291        1
Fe      12.17637877       7.03003556      22.48549291        1
Fe      16.23517170       7.03003556      22.48549291        1
Fe      20.29396462       7.03003556      22.48549291        1
Fe      24.35275754       7.03003556      22.48549291        1
Fe      28.41155047       7.03003556      22.48549291        1
Fe      32.47034339       7.03003556      22.48549291        1
Fe      36.52913632       7.03003556      22.48549291        1
Fe       2.02939646      10.54505334      22.48549291        1
Fe       6.08818939      10.54505334      22.48549291        1
Fe      10.14698231      10.54505334      22.48549291        1
Fe      14.20577523      10.54505334      22.48549291        1
Fe      18.26456816      10.54505334      22.48549291        1
Fe      22.32336108      10.54505334      22.48549291        1
Fe      26.38215401      10.54505334      22.48549291        1
Fe      30.44094693      10.54505334      22.48549291        1
Fe      34.49973985      10.54505334      22.48549291        1
Fe      38.55853278      10.54505334      22.48549291        1
Fe       0.00000000      14.06007112      22.48549291        1
Fe       4.05879292      14.06007112      22.48549291        1
Fe       8.11758585      14.06007112      22.48549291        1
Fe      12.17637877      14.06007112      22.48549291        1
Fe      16.23517170      14.06007112      22.48549291        1
Fe      20.29396462      14.06007112      22.48549291        1
Fe      24.35275754      14.06007112      22.48549291        1
Fe      28.41155047      14.06007112      22.48549291        1
Fe      32.47034339      14.06007112      22.48549291        1
Fe      36.52913632      14.06007112      22.48549291        1
Fe       2.02939646      17.57508890      22.48549291        1
Fe       6.08818939      17.57508890      22.48549291        1
Fe      10.14698231      17.57508890      22.48549291        1
Fe      14.20577523      17.57508890      22.48549291        1
Fe      18.26456816      17.57508890      22.48549291        1
Fe      22.32336108      17.57508890      22.48549291        1
Fe      26.38215401      17.57508890      22.48549291        1
Fe      30.44094693      17.57508890      22.48549291        1
Fe      34.49973985      17.57508890      22.48549291        1
Fe      38.55853278      17.57508890      22.48549291        1
Fe       0.00000000      21.09010669      22.48549291        1
Fe       4.05879292      21.09010669      22.48549291        1
Fe       8.11758585      21.09010669      22.48549291        1
Fe      12.17637877      21.09010669      22.48549291        1
Fe      16.23517170      21.09010669      22.48549291        1
Fe      20.29396462      21.09010669      22.48549291        1
Fe      24.35275754      21.09010669      22.48549291        1
Fe      28.41155047      21.09010669      22.48549291        1
Fe      32.47034339      21.09010669      22.48549291        1
Fe      36.52913632      21.09010669      22.48549291        1
Fe       2.02939646      24.60512447      22.48549291        1
Fe       6.08818939      24.60512447      22.48549291        1
Fe      10.14698231      24.60512447      22.48549291        1
Fe      14.20577523      24.60512447      22.48549291        1
Fe      18.26456816      24.60512447      22.48549291        1
Fe      22.32336108      24.60512447      22.48549291        1
Fe      26.38215401      24.60512447      22.48549291        1
Fe      30.44094693      24.60512447      22.48549291        1
Fe      34.49973985      24.60512447      22.48549291        1
Fe      38.55853278      24.60512447      22.48549291        1
Fe       0.00000000      28.12014225      22.48549291        1
Fe       4.05879292      28.12014225      22.48549291        1
Fe       8.11758585      28.12014225      22.48549291        1
Fe      12.17637877      28.12014225      22.48549291        1
Fe      16.23517170      28.12014225      22.48549291        1
Fe      20.29396462      28.12014225      22.48549291        1
Fe      24.35275754      28.12014225      22.48549291        1
Fe      28.41155047      28.12014225      22.48549291        1
Fe      32.47034339      28.12014225      22.48549291        1
Fe      36.52913632      28.12014225      22.48549291        1
Fe       2.02939646      31.63516003      22.48549291        1
Fe       6.08818939      31.63516003      22.48549291        1
Fe      10.14698231      31.63516003      22.48549291        1
Fe      14.20577523      31.63516003      22.48549291        1
Fe      18.26456816      31.63516003      22.48549291        1
Fe      22.32336108      31.63516003      22.48549291        1
Fe      26.38215401      31.63516003      22.48549291        1
Fe      30.44094693      31.63516003      22.48549291        1
Fe      34.49973985      31.63516003      22.48549291        1
Fe      38.55853278      31.63516003      22.48549291        1
```

**Comment by NikoOinonen** [comment]
Ok, I didn't know that ASE uses the fifth column differently. It's actually a problem, since we could accidentally end up with incorrect electrostatics by using these numbers as charges when we should not. We could make some function that parses and detects that it's in the ASE format and then consequently ignore the fifth column. Alternatively, we could just decide on a standard that says that "the fifth column, if it exists, specifies the charges of the atoms", which would make it incompatible with the ASE-written xyz files. Any thoughts?

The GUI is currently using `loadAtomsLines`, but I will change that.

**Comment by ondrejkrejci** [comment]
As for the ASE - yes the finder of the ASE format is one of few mistake-proofs, that I can imagine.
Other would be to count all those number together and if abs(sum(charge)) >= cut_off ??? (e.g 4+3+2+1 = 10), it would not be usable as charges.  
@ProkopHapala , @yakutovicha and anyone else - some other ideas?

**Comment by NikoOinonen** [comment]
Maybe make that abs(sum(charge)) / len(charges) >= cut_off, to check the average so that it does not depend on the number of atoms.

**Comment by ondrejkrejci** [comment]
My idea was - if chemists are using their oxidation number based integer charges, you normally end up with system <-3,+3> but mostly neutral. But you want to have the option for studying charged systems. At the same time ASE always goes 1+2+3 ... if you have 1x1 cell of 3 layered material.
You can also think like charge max 1 per atom.
I think that we can anytime find a way how to break any simple rule. The point is to cache most of the mistakes here.

**Comment by NikoOinonen** [comment]
I've made the changes to the xyz io functions in this branch: https://github.com/Probe-Particle/ProbeParticleModel/blob/xyz_io/pyProbeParticle/basUtils.py. All of the other xyz read/write function have been replaced with the new ones: loadXYZ and saveXYZ.

With regard to the ASE xyz format, I found that the 'Properties' section in the comment line actually tells what the columns contain. If the column is charges, then it has either the tag 'charge' or 'initial_charges'. I added a function that parses the comment line and returns the lattice vectors as well as the information whether the fifth column contains the charges or not: https://github.com/Probe-Particle/ProbeParticleModel/blob/7f627c4b0d7d05b0066b25cd4760a4c50e1ccb21/pyProbeParticle/basUtils.py#L367

I also added some simple heuristics for detecting false charges:
https://github.com/Probe-Particle/ProbeParticleModel/blob/7f627c4b0d7d05b0066b25cd4760a4c50e1ccb21/pyProbeParticle/basUtils.py#L399-L404

If everything looks right, then I will make pull request after #45 gets merged.

**Comment by yakutovicha** [comment]
Why not just use ASE's `read/write`?

I think they are pretty nice functions and could be used as is in most cases.

Not a big fun of reinventing the wheel. Plus ASE's Atoms object gives a lot of flexibility.

**Comment by yakutovicha** [comment]
Also, concerning the ASE's XYZ format. The comment string describes what information is provided in the file:
```
Properties=species:S:1:pos:R:3:tags:I:1 pbc="T T T"
```
Here is what it means:
- `species:S:1` first column contains species, it is of string type, it has one word
- `pos:R:3` positions, real type, 3 words
- `tags:I:1` tags, integer, one word
-
One could add as many columns as one wants: charges, magnetisation, mass, time on the moon, price of bitcoin, whatever.

I believe we should accept any XYZ file the user provides and act according to the information that is provided. That of course means graceful handling of "incomplete" xyz files.

**Comment by NikoOinonen** [comment]
> Why not just use ASE's `read/write`?
> 
> I think they are pretty nice functions and could be used as is in most cases.
> 
> Not a big fun of reinventing the wheel. Plus ASE's Atoms object gives a lot of flexibility.

I was also wondering about this. I think generally it's better not to add an extra dependency if we don't use it in some significant way. If we use the `Atoms` object here, then it would make sense to use it everywhere, but that would be a much bigger project undertake. In this case, the effort to write our own read/write functions is pretty trivial, since the xyz file format is so simple.

> One could add as many columns as one wants: charges, magnetisation, mass, time on the moon, price of bitcoin, whatever.
> 
> I believe we should accept any XYZ file the user provides and act according to the information that is provided. That of course means graceful handling of "incomplete" xyz files.

Yes, I made it in a kind of naive way assuming that, in addition to the elements and positions, there is only a charges column, if anything. I could change it in a way that it actually takes the relevant column.

What exactly do we mean by "incomplete" xyz files? If there is a line that does not even have the complete elements and positions or the file does not have enough lines for the specified number of atoms, then I think we should just throw an error, since the file does not conform to the file format.

**Comment by ondrejkrejci** [comment]
The reason for it was really not adding any dependency, when there is no heavy need for it. I also do not understand the "incomplete" xyz files. @yakutovicha could you clarify?

**Comment by yakutovicha** [comment]
> I was also wondering about this. I think generally it's better not to add an extra dependency if we don't use it in some significant way. If we use the `Atoms` object here, then it would make sense to use it everywhere, but that would be a much bigger project undertake. In this case, the effort to write our own read/write functions is pretty trivial, since the xyz file format is so simple.

If implementing an extended XYZ format parsing (with the possibility to define different columns) is easier than replacing things with an ASE reader, then it is fine with me 👍 .

However, I do think that in the future, we should embrace the ASE library. Already the capability to read many different formats is a big plus of ASE.

At the same time, unifying the read functions is already a step forward, so I wouldn't stand in your way to add this change.

**What did I mean by incomplete XYZ files?**

To run simulations, one needs the cell, coordinates, and charges. Some of this might be missing, making the XYZ _not complete_. I mean it in the sense of not providing all the information that we might need.

A user might not have provided all information purposefully or mistakenly. So the program should warn, for instance, that charges are missing and the simulations will be run without them.

**Comment by ondrejkrejci** [comment]
Thank you @yakutovicha for clarifying! I see your point now.
Generally we wanted to offer a possibility to use something like _chemdraw_ for the experimentalist to try 1st approximative simulation. This will of course give you no cell, but then you can specify your cell in the **params.ini** file. I would not restrict using the XYZ format (and others) for having always the cell. (charges are absolutely not necessary if one is using more precise hartree potential from DFT). This change will also have problem with back-compability,
I see three possible strategies here:
 - Leave as is and hope that people knows what they are doing.
 - During reading of XYZ without cell, we just simply ask, if one has a properly specified cell in the **params.ini** file and write yes.
  - set:
  
  ```
  gridA 0.0 0.0 0.0
  gridB 0.0 0.0 0.0
  gridC 0.0 0.0 0.0
  ```
  in https://github.com/Probe-Particle/ProbeParticleModel/blob/main/pyProbeParticle/common.py and https://github.com/Probe-Particle/ProbeParticleModel/blob/main/pyProbeParticle/defaults/params.ini 
and then have an assert error if after reading both - XYZ (or other geometry file) and **params.ini** the cell has still 0 volume.

@yakutovicha , @NikoOinonen  and @ProkopHapala - what do you think?

**Comment by NikoOinonen** [comment]
I think most of the time if you just draw some molecule, it's going to be in free space and you don't care about periodicity. We could just leave the grid not specified by default, and if it's also not specified in params.ini or the input file, then we can just choose a grid big enough that it encompasses the whole region specified by `scanMin` and `scanMax`. This is actually what I already did in the `AFMulator` in the OpenCL code when the `lvec` parameter is not specified: https://github.com/Probe-Particle/ProbeParticleModel/blob/3020d0068b2d3fa8e362aa5d55ef444ac6bcf5fe/pyProbeParticle/ocl/AFMulator.py#L398

**Comment by NikoOinonen** [comment]
The main problem of unifying the xyz IO functions was solved, but there is still some discussion left for the handling of missing lattice vectors, so I started a new issue for it at #52.
